libs/scripts.js

```diff
@@ -37,12 +37,30 @@ const requireScript = scriptName => {
 
 /** List all builtin scripts */
 const list = () => {
-  const dir = path.join(__dirname, 'scripts')
-  return fs
-    .readdirSync(dir)
-    .filter(x => /\.js$/.exec(x))
-    .filter(x => !/\.spec\.js$/.exec(x))
-    .map(x => x.replace(/\.js$/, ''))
+  const walkDir = dir => {
+    let files = []
+    const dirContent = fs.readdirSync(dir)
+    dirContent.forEach(fileOrDir => {
+      const pathFileOrDir = path.join(dir, fileOrDir)
+      const stat = fs.statSync(pathFileOrDir)
+      if (stat && stat.isDirectory()) {
+        // Recursively walk in subdir
+        files = files.concat(walkDir(pathFileOrDir))
+      } else {
+        if (/\.js$/.exec(fileOrDir) && !/\.spec\.js$/.exec(fileOrDir)) {
+          let file = fileOrDir.replace(/\.js$/, '')
+          if (path.basename(dir) !== 'scripts') {
+            // Add dir basename for subdirectories
+            file = path.join(path.basename(dir), file)
+          }
+          files.push(file)
+        }
+      }
+    })
+    return files
+  }
```

Comment:
Walkdir could be extracted from this function, I think it would be a little bit clearer as you do not have to worry about variables captured in the closure.

One problem I see is that walkdir also does the filtering for JS files here, and is coupled to the "script" functionality. I think they should be decoupled. One function to recursively walk directories, and then the filtering on /\.js/.

Reply:
Ok, I extracted the walking